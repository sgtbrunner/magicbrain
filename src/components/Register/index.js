import Register from './register.component';

export default Register;
