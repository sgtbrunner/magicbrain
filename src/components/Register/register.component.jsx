import React from 'react';

class Register extends React.Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: '',
    };
  }

  onNameChange = (event) => {
    this.setState({ name: event.target.value });
  };

  onEmailChange = (event) => {
    this.setState({ email: event.target.value.toLowerCase() });
  };

  onPasswordChange = (event) => {
    this.setState({ password: event.target.value });
  };

  handleKeyPress = (event) => {
    if (event.key === 'Enter') {
      this.onRegisterSubmit();
    }
  };

  onRegisterSubmit = () => {
    if (this.state.name && this.state.email && this.state.password) {
      fetch('https://shielded-reaches-78464.herokuapp.com/register', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          name: this.state.name,
          email: this.state.email,
          password: this.state.password,
        }),
      })
        .then((response) => response.json())
        .then((user) => {
          if (user.id) {
            this.props.loadUser(user);
            this.props.onRouteChange('home'); //As Register is a class, it automatically receives props from App.js and (props) are accessed through this.props.
            window.alert(`User ${user.name} succesfully registered!`);
          } else {
            window.alert(user);
            this.props.clearFields();
          }
        });
    } else {
      window.alert('Name, Email and Password fields should be filled in!');
    }
  };

  render() {
    return (
      <article
        className="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 center form smaller"
        onKeyPress={this.handleKeyPress}
      >
        <main className="pa4 black-80">
          <div className="measure">
            <fieldset id="sign_up" className="ba b--transparent ph0 mh0">
              <legend className="f2 fw6 ph0 mh0 noselect">Register</legend>
              <div className="mt3">
                <label className="db fw6 lh-copy f6" htmlFor="email-address">
                  Name
                </label>
                <input
                  className="pa2 input-reset ba bg-white w-100"
                  type="text"
                  name="name"
                  id="name"
                  onChange={this.onNameChange}
                  required
                />
              </div>
              <div className="mt3">
                <label className="db fw6 lh-copy f6" htmlFor="email-address">
                  Email
                </label>
                <input
                  className="pa2 input-reset ba bg-white w-100"
                  type="email"
                  name="email-address"
                  id="email-address"
                  onChange={this.onEmailChange}
                  required
                />
              </div>
              <div className="mv3">
                <label className="db fw6 lh-copy f6" htmlFor="password">
                  Password
                </label>
                <input
                  className="pa2 input-reset ba bg-white w-100"
                  type="password"
                  name="password"
                  id="password"
                  onChange={this.onPasswordChange}
                  required
                />
              </div>
            </fieldset>
            <div className="">
              <input
                className="b ph3 pv2 input-reset ba b--black bg-lightest-blue grow pointer f6 dib"
                type="submit"
                value="Register"
                onClick={this.onRegisterSubmit}
              />
            </div>
          </div>
        </main>
      </article>
    );
  }
}

export default Register;
