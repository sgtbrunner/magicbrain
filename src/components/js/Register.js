import React from 'react';

const Register = ({onRouteChange}) => {
  return (
      <article className="br3 ba b--black-10 mv4 w-100 w-50-m w-25-l mw6 shadow-5 center form smaller">
        <main className="pa4 black-80">
          <form className="measure">
            <fieldset id="sign_up" className="ba b--transparent ph0 mh0">
              <legend className="f1 fw6 ph0 mh0 noselect">Register</legend>
              <div className="mt3">
                <label className="db fw6 lh-copy f6" htmlFor="email-address">Name</label>
                <input 
                  className="pa2 input-reset ba bg-white w-100" 
                  type="text" 
                  name="name"  
                  id="name" 
                />
              </div>
              <div className="mt3">
                <label className="db fw6 lh-copy f6" htmlFor="email-address">Email</label>
                <input 
                  className="pa2 input-reset ba bg-white w-100" 
                  type="email" 
                  name="email-address"  
                  id="email-address" 
                />
              </div>
              <div className="mv3">
                <label className="db fw6 lh-copy f6" htmlFor="password">Password</label>
                <input 
                  className="pa2 input-reset ba bg-white w-100" 
                  type="password" 
                  name="password"  
                  id="password"
                />
              </div>
            </fieldset>
            <div className="">
              <input
                className="b ph3 pv2 input-reset ba b--black bg-lightest-blue grow pointer f6 dib" 
                type="submit" 
                value="Register"
                onClick={() => onRouteChange('signin')}
              />
            </div>
          </form>
        </main>
      </article>
  );
}

export default Register;