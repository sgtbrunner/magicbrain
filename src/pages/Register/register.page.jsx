import React, { useState } from 'react';
import PropTypes from 'prop-types';

import api from '../../utils/api.utils';
import {
  EMAIL,
  PASSWORD,
  INPUT_INITIAL_STATE,
  NAME_ERROR_MESSAGE,
  EMAIL_ERROR_MESSAGE,
  PASSWORD_ERROR_MESSAGE,
  EMAIL_REGEX_KEY,
} from '../../utils/constants.utils';

import FormButton from '../../components/Buttons/FormButton';
import FormInput from '../../components/Inputs/FormInput';

const Register = ({ loadUser }) => {
  const [name, setName] = useState({ ...INPUT_INITIAL_STATE, errorText: NAME_ERROR_MESSAGE });
  const [email, setEmail] = useState({
    ...INPUT_INITIAL_STATE,
    errorText: EMAIL_ERROR_MESSAGE,
  });
  const [password, setPassword] = useState({
    ...INPUT_INITIAL_STATE,
    errorText: PASSWORD_ERROR_MESSAGE,
  });
  const [error, setError] = useState(null);

  const getInput = {
    name,
    email,
    password,
  };

  const getSetInput = {
    name: setName,
    email: setEmail,
    password: setPassword,
  };

  const isValidForm = name.isValid && email.isValid && password.isValid;

  const isFieldValid = (fieldName, value) => {
    const validators = {
      name: value.length > 0,
      email: EMAIL_REGEX_KEY.test(String(value).toLowerCase()),
      password: value.length >= 6,
    };

    return validators[fieldName];
  };

  const onFieldChange = (fieldName, event) => {
    const input = getInput[fieldName];
    const isValid = isFieldValid(fieldName, event.target.value);
    getSetInput[fieldName]({
      ...input,
      value: event.target.value,
      isValid,
      showError: input.showError && !isValid,
    });
  };

  const validateInput = (fieldName) => {
    const input = getInput[fieldName];
    getSetInput[fieldName]({ ...input, showError: !input.isValid });
  };

  const getInputErrorClass = (showError) => (showError ? 'b--red' : '');

  const onRegisterSubmit = (event) => {
    event.preventDefault();
    api
      .registerUser({ name: name.value, email: email.value, password: password.value })
      .then((response) => {
        if (response?.id) {
          loadUser(response);
          localStorage.setItem('user', JSON.stringify(response));
        } else {
          setError(response.error);
        }
      })
      .catch(() => setError('Something went wrong. Please try again later.'));
  };

  return (
    <form
      className="br3 ba b--black-10 mv4 w-100 w-50-m mw6 shadow-5 center smaller"
      onSubmit={onRegisterSubmit}
    >
      <main className="pv3 ph4 black-80">
        <fieldset id="sign-in" className="ba b--transparent ph0 mh0">
          <h2 className="f2 ph0 ma0 noselect">Register</h2>
          {error && (
            <p className="red f6 mv3 pv3 right-0 left-0 b bg-light-pink ba b--red br2">{error}</p>
          )}
          <FormInput
            name="name"
            errorData={{ showError: name.showError, errorText: name.errorText }}
            errorClass={getInputErrorClass(name.showError)}
            onChange={onFieldChange}
            onBlur={validateInput}
            isTopInput
          />
          <FormInput
            name={EMAIL}
            type={EMAIL}
            errorData={{ showError: email.showError, errorText: email.errorText }}
            errorClass={getInputErrorClass(email.showError)}
            onChange={onFieldChange}
            onBlur={validateInput}
          />
          <FormInput
            name={PASSWORD}
            type={PASSWORD}
            errorData={{ showError: password.showError, errorText: password.errorText }}
            errorClass={getInputErrorClass(password.showError)}
            onChange={onFieldChange}
            onBlur={validateInput}
          />
        </fieldset>
        <FormButton isValidForm={isValidForm} disabled={!isValidForm}>
          Register
        </FormButton>
      </main>
    </form>
  );
};

Register.propTypes = {
  loadUser: PropTypes.func.isRequired,
};

export default Register;
