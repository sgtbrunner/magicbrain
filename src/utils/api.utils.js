const { REACT_APP_API_ENDPOINT } = process.env;

const fetchServer = ({
  path,
  method = 'POST',
  headers = { 'Content-Type': 'application/json' },
  body,
}) =>
  fetch(`${REACT_APP_API_ENDPOINT}${path}`, {
    method,
    headers,
    body: JSON.stringify({ ...body }),
  });

export default {
  signInUser: ({ email, password }) =>
    fetchServer({ path: '/signin', body: { email, password } }).then((response) => response.json()),

  registerUser: ({ name, email, password }) =>
    fetchServer({ path: '/register', body: { name, email, password } }).then((response) =>
      response.json()
    ),

  getImageBoundary: ({ input }) =>
    fetchServer({ path: '/imageurl', body: { input } }).then((response) => response.json()),

  updateImageCount: ({ id }) =>
    fetchServer({ path: '/image', method: 'PUT', body: { id } }).then((response) =>
      response.json()
    ),
};
