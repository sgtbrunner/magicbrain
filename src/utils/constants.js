export const PARTICLE_OPTIONS = {
  particles: {
    number: {
      value: 50,
      density: {
        enable: true,
        value_area: 500,
      },
    },
  },
};

export const { REACT_APP_API_ENDPOINT } = process.env;
