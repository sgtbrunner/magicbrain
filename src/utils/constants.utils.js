const PARTICLE_OPTIONS = {
  particles: {
    number: {
      value: 50,
      density: {
        enable: true,
        value_area: 500,
      },
    },
  },
};

export default PARTICLE_OPTIONS;
