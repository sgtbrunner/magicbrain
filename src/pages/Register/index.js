import Register from './register.page';

export default Register;
