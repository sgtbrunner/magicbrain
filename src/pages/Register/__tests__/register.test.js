import { screen } from '@testing-library/react';

import { capitalizeFirstLetter } from '../../../utils/functions.utils';
import { mountComponent } from '../../../utils/test.utils';
import Register from '..';

const BUTTON = 'button';
const EMAIL = 'email';
const FORM_INPUT = 'form-input';
const HEADING = 'heading';
const PASSWORD = 'password';
const REGISTER = 'Register';

const mountRegisterPage = () =>
  mountComponent({
    component: Register,
    props: {
      loadUser: jest.fn(),
    },
  });

describe('Register page', () => {
  it(`should match snapshot`, () => {
    const { container } = mountRegisterPage();

    expect(container).toMatchSnapshot();
  });

  it(`should render 'Register' label and button`, () => {
    mountRegisterPage();

    const label = screen.getByRole(HEADING, { name: REGISTER });
    const button = screen.getByRole(BUTTON, { name: REGISTER });

    expect(label).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  it(`should render three (3) input fields (name, email and password)`, () => {
    mountRegisterPage();

    const inputFields = screen.getAllByTestId(FORM_INPUT);
    expect(inputFields).toHaveLength(3);

    const nameInput = screen.getByLabelText(capitalizeFirstLetter('name'));
    expect(nameInput).toBeInTheDocument();

    const emailInput = screen.getByLabelText(capitalizeFirstLetter(EMAIL));
    expect(emailInput).toBeInTheDocument();

    const passwordInput = screen.getByLabelText(capitalizeFirstLetter(PASSWORD));
    expect(passwordInput).toBeInTheDocument();
  });

  it(`should have submit button disabled if input fields are invalid`, () => {
    mountRegisterPage();

    const button = screen.getByRole(BUTTON, { name: REGISTER });
    expect(button).toBeDisabled();
  });
});
