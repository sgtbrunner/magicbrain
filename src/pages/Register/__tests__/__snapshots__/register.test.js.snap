// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register page should match snapshot 1`] = `
<div>
  <form
    class="br3 ba b--black-10 mv4 mw6 shadow-5 center smaller"
  >
    <main
      class="pv3 ph4 black-80"
    >
      <fieldset
        class="ba b--transparent ph0 mh0"
        id="sign-in"
      >
        <h2
          class="f2 ph0 ma0 noselect"
        >
          Register
        </h2>
        <div
          class=""
          data-testid="form-input"
        >
          <label
            class="db fw6 lh-copy mv2"
            for="name"
          >
            Name
            <input
              class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
              id="name"
              name="name"
              type="text"
            />
          </label>
        </div>
        <div
          class="mt4"
          data-testid="form-input"
        >
          <label
            class="db fw6 lh-copy mv2"
            for="email"
          >
            Email
            <input
              class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
              id="email"
              name="email"
              type="email"
            />
          </label>
        </div>
        <div
          class="mt4"
          data-testid="form-input"
        >
          <label
            class="db fw6 lh-copy mv2"
            for="password"
          >
            Password
            <input
              class="pa2 mb0 ba bg-white w-100 br2 () => showError ? 'b--red' : ''"
              id="password"
              name="password"
              type="password"
            />
          </label>
        </div>
      </fieldset>
      <div
        class="flex flex-column items-center"
      >
        <button
          class="ph3 pv2 mt3 ba b--black f6 dib "
          disabled=""
          type="submit"
        >
          Register
        </button>
      </div>
    </main>
  </form>
</div>
`;
